Signup: Prevent the site step in signup from being submitted twice

scruffian opened the pull request:
This pull request prevents the site step in the signup flow from submitting to the API twice. We were seeing some instances of this form being submitted twice which was causing some strange behaviour through the API. This PR seeks to address that by making it impossible to submit this step more than once.
 
#### Testing instructions

This is hard to test.

1. Run `git checkout fix/double-submit-site-step` and start your server
2. Open the [`Free trials signup` page](http://calypso.dev:3000/start/free-trials) as a logged in user
3. Go to the site step
4. Try to submit the site step twice. You have to be a logged in user so the request is made right away.

You can see the requests in the network tab to /sites/new - there are two different requests - one with a `validate` property `true` - these won't create a site, only validate it. We can make as many of these as we want. However when the `validate` property is `false` we will create the site. There should only be one request of this type.

Some things I tried to replicate the problem, or cause other problems are:
- Going back to the step after submitting it and seeing if I could submit it again
- Hit the enter button a lot of times to do lots of submits

#### Reviews

- [x] Code
- [x] Product

spncrb commented:
With some help from @drewblaisdell  I was able to test this locally...I wasn't able to get more than 1 validation:false request from multiple attempts to click the button, and going back...the next screen actually loads very quickly so getting more than ~5 clicks is pretty difficult and when you return to the screen it has the "site already created" message.

:+1: 

PS. I wasn't that familiar with the Network tab, hence asking for help from Drew. In the future, maybe a screenshot or slightly more detailed testing instructions would benefit a less-development-trained tester like myself

scruffian commented:
> PS. I wasn't that familiar with the Network tab, hence asking for help from Drew. In the future, maybe a screenshot or slightly more detailed testing instructions would benefit a less-development-trained tester like myself

Good feedback, thanks.

drewblaisdell commented:
Were you ever able to replicate the bug? It isn't clear to me why it would be possible for `SignupFlowController` to call the site step's action multiple times, but if it is, this is probably something we should fix in that module.

scruffian commented:
@drewblaisdell - I can replicate the issue - you just go back the site step and submit it again (in master).

I added a commit to take care of this in the framework, by making it impossible to update steps that are `processing` or `completed` but I am not sure the place I added it is the best, or if there is already an attempt to solve this, or if this will cause other issues...

drewblaisdell commented:
@scruffian I started to write a test for this case, but realized that there are legitimate cases for submitting a step multiple times, like in the case of the themes step. There aren't enough cases of this that I think this should be an additional property in the config to create another branch of behavior when submitting a step. Your solution from the first commit is ideal. Sorry about that.

I removed 684156d5a751efb6061a90528a0c8d63699c027c. @spncrb gave the other commits a :+1:, and I tested the PR again, so this is ready to merge.
